### ISSUE TITLE
Support "auto reaping"

### ISSUE THREAD

**Issue by merrellb** [python-trio/trio#135]
The tutorial and examples don't go into much detail on reaping best practices.  As clients connect and disconnect from long-running nurseries, should we be reaping to prevent the monitor queue and zombie frozenset from growing forever?  Perhaps an autoreap option when creating the nursery (disabling the monitor and zombie attributes) or taking an async function argument for minimal cleanup.

```
async def autoreap(nursery, reap_fn=None):
    while True:
        batch = await nursery.monitor.get_batch()
        for task in batch:
            if reap_fn:
                await reap_fn(task)
            nursery.reap(task)
```

**Comment by njsmith** [comment]
Ah, I think I see the problem... The docs are not as clear as they could be about an important issue :-)

The intended way to use nurseries (unless you're actually implementing a custom supervisor) is that you spawn some tasks and then sit parked in the nursery `__aexit__`. The task that holds the nursery shouldn't be doing other work itself; it should all be delegated to children. ("Parenting is a full time job.")

I think this needs to be emphasized more in the docs, and in particular the tutorial... I'm seeing now that it's really not obvious (esp. since the call to `__aexit__` is literally invisible!). But even though from the user point of view it looks like the parent task isn't doing anything, in fact it's performing the critical work of watching for child tasks to exit, reaping them, and responding to any unexpected exceptions that bubble up from the children. (I think this is also why you were having trouble with `KeyboardExit`not working.)

Basically the nursery `__aexit__` **is** your autoreap function and more :-). (In particular, the snippet you show silently discards any unexpected exceptions; if you look at the code that implements nurseries then like 90% of the complications are from dealing with exceptions. Exceptions are super annoying.)

Does that make sense?

**Comment by merrellb** [comment]
Yeah, the invisible ```__aexit__``` is both quite elegant and invisible enough to be potentially confusing :-) .  I sometimes find myself wishing for something a bit more explicit like requiring an ```await nursery.done()``` at the end of a nursery scope (which would also give something to wrap in a try/except block for KeyboardInterrupt).

In this case, I am thinking about long running servers where a nursery could be going for days or months.  Each connection will have multiple tasks starting and ending which will need to be managed within the nursery scope.   Some simple tests I've tried suggests that ```__aexit__``` doesn't perform any reaping until the nursery scope actually ends.  With a long running server, how do we prevent zombies from growing indefinitely without some explicit reaping task?
